config/option.go

```diff
@@ -34,32 +37,174 @@ func getTypeName(t uint8) string {
-	DefaultValue    interface{}
-
-	ExternalOptType string
+	// DefaultValue holds the default value of the option. Note that
+	// this value can be overwritten during runtime (see activeDefaultValue
+	// and activeFallbackValue).
+	DefaultValue interface{}
+	// ValidationRegex may contain a regular expression used to validate
+	// the value of option. If the option type is set to OptTypeStringArray
+	// the validation regex is applied to all entries of the string slice.
+	// Note that it is recommended to keep the validation regex simple so
+	// it can also be used in other languages (mainly JavaScript) to provide
+	// a better user-experience by pre-validating the expression.
 	ValidationRegex string
+	// PossibleValues may be set to a slice of values that are allowed
+	// for this configuration setting. Note that PossibleValues makes most
+	// sense when ExternalOptType is set to HintOneOf
+	PossibleValues []PossibleValue `json:",omitempty"`
+	// Annotations adds additional annotations to the configuration options.
+	// See documentation of Annotations for more information.
+	Annotations Annotations
 
 	activeValue         *valueCache // runtime value (loaded from config file or set by user)
 	activeDefaultValue  *valueCache // runtime default value (may be set internally)
 	activeFallbackValue *valueCache // default value from option registration
 	compiledRegex       *regexp.Regexp
 }
 
+// AddAnnotation adds the annotation key to option if it's not already set.
+func (option *Option) AddAnnotation(key string, value interface{}) {
+	option.Lock()
+	defer option.Unlock()
+
+	option.AddAnnotationUnsafe(key, value)
+}
+
+// AddAnnotationUnsafe is like AddAnnotation but does not lock the option.
+func (option *Option) AddAnnotationUnsafe(key string, value interface{}) {
+	if option.Annotations == nil {
+		option.Annotations = make(Annotations)
+	}
+
+	if _, ok := option.Annotations[key]; ok {
+		return
+	}
+	option.Annotations[key] = value
+}
+
+// SetAnnotation sets the value of the annotation key overwritting an
+// existing value if required.
+func (option *Option) SetAnnotation(key string, value interface{}) {
+	option.Lock()
+	defer option.Unlock()
+
+	option.SetAnnotationUnsafe(key, value)
+}
+
+// SetAnnotationUnsafe is like SetAnnotation but does not lock the option.
+func (option *Option) SetAnnotationUnsafe(key string, value interface{}) {
+	if option.Annotations == nil {
+		option.Annotations = make(Annotations)
+	}
+	option.Annotations[key] = value
+}
+
+// GetAnnotation returns the value of the annotation key
+func (option *Option) GetAnnotation(key string) (interface{}, bool) {
+	option.Lock()
+	defer option.Unlock()
+	return option.GetAnnotationUnsafe(key)
+}
+
+// GetAnnotationUnsafe is like GetAnnotation but does not lock the option.
+func (option *Option) GetAnnotationUnsafe(key string) (interface{}, bool) {
+	if option.Annotations == nil {
+		return nil, false
+	}
+	val, ok := option.Annotations[key]
+	return val, ok
+}
```

Comment:
Is there a reason for exporting the `Unsafe` method variants?
I'd like to follow a convention where exported functions are safe to use in regards to locking, but internal functions usually aren't.